### PR TITLE
Expose sequence number and primary terms in search responses

### PR DESCRIPTION
**Pull request by bleskes**
Users may require the sequence number and primary terms to perform optimistic concurrency control operations. Currently, you can get the sequence number via the `docvalues_fields` API but the primary term is not accessible because it is maintained by the `SeqNoFieldMapper` and the infrastructure can't find it. 

This PR adds a dedicated sub fetch phase to return both numbers that is connected to a new `seq_no_primary_term` parameter.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/search/SearchHit.java [hunk 1]
```diff
@@ -587,6 +618,8 @@ public static SearchHit createFromMap(Map<String, Object> values) {
         }
         searchHit.score(get(Fields._SCORE, values, DEFAULT_SCORE));
         searchHit.version(get(Fields._VERSION, values, -1L));
+        searchHit.setSeqNo(get(Fields._SEQ_NO, values, SequenceNumbers.UNASSIGNED_SEQ_NO));
```
Comment (jimczi) [inline server/src/main/java/org/elasticsearch/search/SearchHit.java]: You also need to declare the field in `SearchHit#declareInnerHitsParseFields` ?

File: server/src/main/java/org/elasticsearch/search/SearchHit.java [hunk 2]
```diff
@@ -587,6 +618,8 @@ public static SearchHit createFromMap(Map<String, Object> values) {
...
+        searchHit.setPrimaryTerm(get(Fields._PRIMARY_TERM, values, SequenceNumbers.UNASSIGNED_PRIMARY_TERM));
```
Comment (jimczi) [inline server/src/main/java/org/elasticsearch/search/SearchHit.java]: Same here ?

File: server/src/main/java/org/elasticsearch/search/SearchHit.java [hunk 3]
```diff
@@ -167,6 +171,25 @@ public long getVersion() {
...
 
+
+    public void setSeqNo(long seqNo) {
+        this.seqNo = seqNo;
+    }
+
+    public void setPrimaryTerm(long primaryTerm) {
+        this.primaryTerm = primaryTerm;
+    }
+
```
Comment (jimczi) [inline server/src/main/java/org/elasticsearch/search/SearchHit.java]: Can you add these fields in SearchHitTests#createTestItem to test all types of (de)serialization ?

**Review by jimczi — CHANGES_REQUESTED** [review]
The change looks good, I left some comments regarding xcontent parsing and tests.

File: server/src/main/java/org/elasticsearch/search/SearchHit.java (hunk 2 above)
Comment (bleskes) [reply server/src/main/java/org/elasticsearch/search/SearchHit.java]: done. What tests am I missing that were supposed to catch it?

Comment (jimczi) [reply server/src/main/java/org/elasticsearch/search/SearchHit.java]: `SearchHitTests` and `SearchHitsTests`, they are supposed to test the (de)serialization of random search hits using xcontent and streamable but you need to fill these fields in the `createTestItem` function to ensure that they are tested.

**Review by jimczi — APPROVED** [review]
LGTM

**Review by jimczi — APPROVED** [review]
LGTM again
Thanks for iterating @bleskes

File: server/src/main/java/org/elasticsearch/search/SearchHit.java [hunk 4]
```diff
@@ -167,6 +171,25 @@ public long getVersion() {
...
+    /** returns the sequence number of the last modification to the document */
+    public long getSeqNo() {
+        return this.seqNo;
+    }
+
+    /** returns the primary term of the last modification to the document */
+    public long getPrimaryTerm() {
+        return this.primaryTerm;
+    }
```
Comment (jpountz) [inline server/src/main/java/org/elasticsearch/search/SearchHit.java]: maybe document default return values, if seq_no_primary_term is not set?

**Review by jpountz — APPROVED** [review]

File: server/src/main/java/org/elasticsearch/search/SearchHit.java (hunk 4 above)
Comment (bleskes) [reply server/src/main/java/org/elasticsearch/search/SearchHit.java]: +1

**Comment by bleskes** [comment]
@elasticmachine please run elasticsearch-ci/2

**Comment by bleskes** [comment]
@elasticmachine please run elasticsearch-ci/2

**Pull request merged**